File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func subtreeWithAllDeepest(root *TreeNode) *TreeNode {
    maxDepth := 0
    depth := map[*TreeNode]int{}

    // dfs find max depth
    var dfs func(node *TreeNode, d int)
    dfs = func(node *TreeNode, d int) {
        if node == nil {
            return
        }

        maxDepth = max(maxDepth, d)
        depth[node] = d

        dfs(node.Left, d+1)
        dfs(node.Right, d+1)
    }

    dfs(root, 1)

    var solve func(node *TreeNode) *TreeNode
    solve = func(node *TreeNode) *TreeNode {
        if node == nil || depth[node] == maxDepth {
            return node
        }

        left, right := solve(node.Left), solve(node.Right) 
        if left != nil && right != nil {
            return node
        } else if left != nil {
            return left
        } else if right != nil {
            return right
        }
        return nil
    }

    return solve(root)
}

// binary tree, dfs, hash map, tree
// time: O(n)
// space: O(n)

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        depth = {None: -1}
        max_depth = 0
        def dfs(node, d, parent = None):
            nonlocal max_depth
            if node:
                max_depth = max(max_depth, d)
                depth[node] = depth[parent] + 1
                dfs(node.left, d + 1,  node)
                dfs(node.right, d + 1, node)
        
        dfs(root, 0)

        def solve(node):
            if not node or depth.get(node, None) == max_depth:
                return node
            left, right = solve(node.left), solve(node.right)
            if left and right:
                return node
            elif left:
                return left
            elif right:
                return right
            else:
                return None
        
        return solve(root)

# binary tree, dfs, hash map, tree
# time: O(n)
# space: O(n)